core/client/fs/src/main/java/alluxio/client/block/stream/BlockWorkerDataWriter.java

```diff
@@ -0,0 +1,159 @@
+
+import java.io.IOException;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A data writer that issues from a client embedded in a worker to write data to the worker directly
+ * via internal communication channel.
+ *
+ * This data writer is similar to write to local worker storage via {@link GrpcDataWriter} except
+ * that all communication with the local worker is via internal method call instead of external RPC
+ * frameworks.
+ */
+@NotThreadSafe
+public final class BlockWorkerDataWriter implements DataWriter {
+  private final long mBlockId;
+  private final BlockWriter mBlockWriter;
+  private final BlockWorker mBlockWorker;
+  private final int mChunkSize;
+  private final OutStreamOptions mOptions;
+  private final long mSessionId;
+  private final long mBufferSize;
+  private long mReservedBytes;
+
+  /**
+   * Creates an instance of {@link BlockWorkerDataWriter}.
+   *
+   * @param context the file system context
+   * @param blockId the block ID
+   * @param blockSize the block size in bytes
+   * @param options the output stream options
+   * @return the {@link BlockWorkerDataWriter} created
+   */
+  public static BlockWorkerDataWriter create(final FileSystemContext context, long blockId,
+      long blockSize, OutStreamOptions options) throws IOException {
+    AlluxioConfiguration conf = context.getClusterConf();
+    int chunkSize = (int) conf.getBytes(PropertyKey.USER_LOCAL_WRITER_CHUNK_SIZE_BYTES);
+    long reservedBytes = Math.min(blockSize, conf.getBytes(PropertyKey.USER_FILE_RESERVED_BYTES));
+    BlockWorker blockWorker = context.getInternalBlockWorker();
+    Preconditions.checkNotNull(blockWorker, "blockWorker");
+    long sessionId = SessionIdUtils.createSessionId();
+    try {
+      blockWorker.createBlock(sessionId, blockId, options.getWriteTier(), options.getMediumType(),
+          reservedBytes);
+      BlockWriter blockWriter = blockWorker.getBlockWriter(sessionId, blockId);
+      return new BlockWorkerDataWriter(sessionId, blockId, options, blockWriter, blockWorker,
+          chunkSize, reservedBytes, conf);
+    } catch (BlockAlreadyExistsException | WorkerOutOfSpaceException | BlockDoesNotExistException
+        | InvalidWorkerStateException e) {
+      throw new IOException(e);
+    }
+  }
+
+  @Override
+  public long pos() {
+    return mBlockWriter.getPosition();
+  }
+
+  @Override
+  public int chunkSize() {
+    return mChunkSize;
+  }
+
+  @Override
+  public void writeChunk(final ByteBuf buf) throws IOException {
+    if (mReservedBytes < pos() + buf.readableBytes()) {
+      try {
+        long bytesToReserve = Math.max(mBufferSize, pos() + buf.readableBytes() - mReservedBytes);
+        // Allocate enough space in the existing temporary block for the write.
+        mBlockWorker.requestSpace(mSessionId, mBlockId, bytesToReserve);
+      } catch (Exception e) {
+        throw new IOException(e);
+      }
+    }
+    long append = mBlockWriter.append(buf);
```

Comment:
If `writeChunk` throws an exception, will the DataWriter guarantee to be cancelled?